scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/context/TaskContextVariableExtractor.java

```diff
@@ -122,11 +122,17 @@
 
     public Map<String, Serializable> extractScopeVariables(TaskContext taskContext) {
         Map<String, Serializable> variables = new HashMap<>();
-
+        Map<String, Serializable> previousVariables = null;
+        try {
+            previousVariables = extractVariables(taskContext, false);
+        } catch (IOException | ClassNotFoundException e) {
+            logger.error("Unable to extract variables", e);
+        }
         // variables from task definition
         if (taskContext.getInitializer().getTaskVariables() != null) {
             for (TaskVariable taskVariable : taskContext.getInitializer().getTaskVariables().values()) {
```

Comment:
You can put this for loop into a private method called something expressive like mergeInheritedVariables or so. 

Possibly, you could even create a new class for this. A ....VariablesMerger.... class, I think the merging is not necessarily part of the extracting. (As you wish) 

When you do that, you are removing the depth of the nested ifs and for.

It would look like this:
1) If the task variables exists, 2) merge (or find a better term for it)  them (keep the merged result in a map). That would hide the logic of the actual merge and give it an expressive name (the if clause is not very expressive at first sight)